src/Features/LanguageServer/Protocol/ExternalAccess/Razor/GetProjectContextsHandler.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.ExternalAccess.Razor;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(GetProjectContextsHandler)), Shared]
+[Method(GetProjectContextsName)]
+internal class GetProjectContextsHandler : ILspServiceDocumentRequestHandler<ProjectContextsParams, ProjectContextList?>
+{
+    private const string GetProjectContextsName = "roslyn/getProjectContexts";
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public GetProjectContextsHandler()
+    {
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(ProjectContextsParams request)
+        => request.TextDocument;
+
+    public Task<ProjectContextList?> HandleRequestAsync(ProjectContextsParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(context.Workspace);
```

Comment:
Copying logic was the easiest path forward, but it does mean that it becomes more difficult to maintain. I'm happy to move [shared code](https://github.com/dotnet/roslyn/blob/c0e4fb4f07e0ca95572ea76ebb65ee541d889640/src/Features/LanguageServer/Protocol/Handler/ProjectContext/GetTextDocumentWithContextHandler.cs#L33) to a static location as needed.

Reply:
I'll leave it up to Roslyn folks to make the final determination, but it seems logical to me that there could just be a static method that produces a `VSProjectContextList` given a document Uri, and then this code could just post-process that list and call `ProtocolConversions.ProjectContextToProjectId` to pull out the project and make the map.